File: src/qov-types.ts
// QOV Format Types based on qov-specification.md

export interface QovHeader {
  magic: string;           // "qovf"
  version: number;         // 0x01, 0x02, or 0x03 (lossy)
  flags: number;           // Feature flags bitfield
  width: number;           // 1-65535
  height: number;          // 1-65535
  frameRateNum: number;    // Frame rate numerator
  frameRateDen: number;    // Frame rate denominator
  totalFrames: number;     // 0 = unknown/streaming
  audioChannels: number;   // 0-8
  audioRate: number;       // Sample rate 0-16777215
  colorspace: number;      // Color space identifier
  // Lossy extension (version 0x03)
  quality?: number;        // Quality level 0-100 (100 = near-lossless)
  yQuantBase?: number;     // Y plane base quantization (1-64)
  uvQuantBase?: number;    // UV plane base quantization (1-64)
  temporalThresh?: number; // Temporal similarity threshold (0-32)
  dctQpBase?: number;      // Base QP for DCT blocks (0-51)
}

// Header flags
export const QOV_FLAG_HAS_ALPHA = 0x01;
export const QOV_FLAG_HAS_MOTION = 0x02;
export const QOV_FLAG_HAS_INDEX = 0x04;
export const QOV_FLAG_HAS_BFRAMES = 0x08;
export const QOV_FLAG_ENHANCED_COMP = 0x10;
export const QOV_FLAG_LOSSY_MODE = 0x20;  // Lossy encoding enabled
export const QOV_FLAG_DCT_ENABLED = 0x40; // DCT block encoding available

// Version constants
export const QOV_VERSION_ORIGINAL = 0x01;    // 16-bit chunk sizes, lossless
export const QOV_VERSION_EXTENDED = 0x02;    // 32-bit chunk sizes, lossless
export const QOV_VERSION_LOSSY = 0x03;       // 32-bit chunk sizes, lossy support

// Colorspace values
export const QOV_COLORSPACE_SRGB = 0x00;
export const QOV_COLORSPACE_SRGBA = 0x01;
export const QOV_COLORSPACE_LINEAR = 0x02;
export const QOV_COLORSPACE_LINEAR_A = 0x03;
export const QOV_COLORSPACE_YUV420 = 0x10;
export const QOV_COLORSPACE_YUV422 = 0x11;
export const QOV_COLORSPACE_YUV444 = 0x12;
export const QOV_COLORSPACE_YUVA420 = 0x13;

// Chunk types
export const QOV_CHUNK_SYNC = 0x00;
export const QOV_CHUNK_KEYFRAME = 0x01;
export const QOV_CHUNK_PFRAME = 0x02;
export const QOV_CHUNK_BFRAME = 0x03;
export const QOV_CHUNK_AUDIO = 0x10;
export const QOV_CHUNK_INDEX = 0xF0;
export const QOV_CHUNK_END = 0xFF;

// Chunk flags
export const QOV_CHUNK_FLAG_YUV = 0x01;         // bit 0: YUV mode
export const QOV_CHUNK_FLAG_MOTION = 0x02;      // bit 1: motion vectors
export const QOV_CHUNK_FLAG_COMPRESSED = 0x10;  // bit 4: LZ4 compressed
export const QOV_CHUNK_FLAG_DCT_BLOCKS = 0x20;  // bit 5: DCT blocks

// Compression types (bits 4-5 of chunk flags)
export const QOV_COMPRESSION_NONE = 0x00;
export const QOV_COMPRESSION_LZ4 = 0x10;

// DCT Opcodes (Version 0x03)
export const QOV_OP_DCT_Y = 0x50;
export const QOV_OP_DCT_UV = 0x51;
export const QOV_OP_DCT_SKIP = 0x52;
export const QOV_OP_DCT_ZERO = 0x53;

// Chunk header
export interface QovChunkHeader {
  chunkType: number;
  chunkFlags: number;
  chunkSize: number;
  timestamp: number;          // microseconds
  uncompressedSize?: number;  // only present if compressed
}

// Index entry for seeking
export interface QovIndexEntry {
  frameNum: number;
  fileOffset: bigint;
  timestamp: number;
}

// RGBA pixel
export interface QovRGBA {
  r: number;
  g: number;
  b: number;
  a: number;
}

// Decoded frame
export interface QovFrame {
  pixels: Uint8ClampedArray;  // RGBA data
  timestamp: number;          // microseconds
  isKeyframe: boolean;
  frameNumber: number;
}

// File statistics for player UI
export interface QovFileStats {
  header: QovHeader;
  fileSize: number;
  chunks: QovChunkInfo[];
  keyframeIndices: number[];
  indexTable: QovIndexEntry[];
  duration: number;          // microseconds
}

export interface QovChunkInfo {
  type: number;
  typeName: string;
  offset: number;
  size: number;
  timestamp: number;
  isKeyframe: boolean;
  isCompressed?: boolean;
  uncompressedSize?: number;
}

export function getChunkTypeName(type: number | undefined): string {
  if (type === undefined || type === null) {
    return 'INVALID(undefined)';
  }
  switch (type) {
    case QOV_CHUNK_SYNC: return 'SYNC';
    case QOV_CHUNK_KEYFRAME: return 'KEYFRAME';
    case QOV_CHUNK_PFRAME: return 'PFRAME';
    case QOV_CHUNK_BFRAME: return 'BFRAME';
    case QOV_CHUNK_AUDIO: return 'AUDIO';
    case QOV_CHUNK_INDEX: return 'INDEX';
    case QOV_CHUNK_END: return 'END';
    default: return `UNKNOWN(0x${type.toString(16)})`;
  }
}

// Lossy encoding parameters
export interface LossyParams {
  yQuant: number;        // Y plane quantization step (1-64)
  uvQuant: number;       // UV plane quantization step (1-64)
  temporalThresh: number; // Temporal similarity threshold (0-32)
  dctQp: number;         // DCT quantization parameter (0-51)
}

// Derive lossy parameters from quality level (0-100)
export function deriveLossyParams(quality: number): LossyParams {
  // Clamp quality to valid range
  quality = Math.max(0, Math.min(100, quality));

  // Y plane quantization (preserve luminance detail)
  const yQuant = Math.max(1, Math.min(64, 1 + Math.floor((100 - quality) / 8)));

  // UV plane quantization (can be more aggressive)
  const uvQuant = Math.max(1, Math.min(64, 2 + Math.floor((100 - quality) / 4)));

  // Temporal similarity threshold for P-frames
  const temporalThresh = Math.max(0, Math.min(32, Math.floor((100 - quality) / 12)));

  // DCT quantization parameter (if DCT enabled)
  const dctQp = Math.max(0, Math.min(51, 51 - Math.floor(quality * 51 / 100)));

  return { yQuant, uvQuant, temporalThresh, dctQp };
}

// Get quality level name
export function getQualityName(quality: number): string {
  if (quality >= 95) return 'Near-lossless';
  if (quality >= 85) return 'High';
  if (quality >= 70) return 'Medium-High';
  if (quality >= 50) return 'Medium';
  if (quality >= 30) return 'Low';
  return 'Very Low';
}

File: csharp_qov/QovLibrary/QovTypes.cs
namespace QovLibrary;

/// <summary>
/// QOV (Quite OK Video) format types and constants.
/// Based on QOV specification version 1.0.
/// </summary>
public static partial class QovTypes
{
    public const string Magic = "qovf";
    public const byte Version1 = 0x01;
    public const byte Version2 = 0x02;

    public const byte FlagHasAlpha = 0x01;
    public const byte FlagHasMotion = 0x02;
    public const byte FlagHasIndex = 0x04;
    public const byte FlagHasBFrames = 0x08;
    public const byte FlagEnhancedComp = 0x10;

    public const byte ChunkFlagYuv = 0x01;
    public const byte ChunkFlagMotion = 0x02;
    public const byte ChunkFlagCompressed = 0x10;

    public const byte ChunkTypeSync = 0x00;
    public const byte ChunkTypeKeyframe = 0x01;
    public const byte ChunkTypePframe = 0x02;
    public const byte ChunkTypeBframe = 0x03;
    public const byte ChunkTypeAudio = 0x10;
    public const byte ChunkTypeIndex = 0xF0;
    public const byte ChunkTypeEnd = 0xFF;

    public const byte ColorspaceSrgb = 0x00;
    public const byte ColorspaceSrgba = 0x01;
    public const byte ColorspaceLinear = 0x02;
    public const byte ColorspaceLinearA = 0x03;
    public const byte ColorspaceYuv420 = 0x10;
    public const byte ColorspaceYuv422 = 0x11;
    public const byte ColorspaceYuv444 = 0x12;
    public const byte ColorspaceYuva420 = 0x13;

    public const byte RunMaxCount = 62;
    public const byte SkipMaxCount = 62;

    public static string GetChunkTypeName(byte chunkType) => chunkType switch
    {
        ChunkTypeSync => "SYNC",
        ChunkTypeKeyframe => "KEYFRAME",
        ChunkTypePframe => "PFRAME",
        ChunkTypeBframe => "BFRAME",
        ChunkTypeAudio => "AUDIO",
        ChunkTypeIndex => "INDEX",
        ChunkTypeEnd => "END",
        _ => $"UNKNOWN(0x{chunkType:X2})"
    };
}

public readonly struct QovHeader
{
    public string Magic { get; init; }
    public byte Version { get; init; }
    public byte Flags { get; init; }
    public ushort Width { get; init; }
    public ushort Height { get; init; }
    public ushort FrameRateNum { get; init; }
    public ushort FrameRateDen { get; init; }
    public uint TotalFrames { get; init; }
    public byte AudioChannels { get; init; }
    public uint AudioRate { get; init; }
    public byte Colorspace { get; init; }

    public QovHeader(byte flags, ushort width, ushort height, ushort frameRateNum = 30, ushort frameRateDen = 1,
        byte colorspace = QovTypes.ColorspaceSrgb, byte audioChannels = 0, uint audioRate = 0,uint totalFrames=0)
    {
        Magic = QovTypes.Magic;
        Version = QovTypes.Version2;
        Flags = flags;
        Width = width;
        Height = height;
        FrameRateNum = frameRateNum;
        FrameRateDen = frameRateDen;
        TotalFrames = totalFrames;
        AudioChannels = audioChannels;
        AudioRate = audioRate;
        Colorspace = colorspace;
    }
}

public readonly struct QovChunk
{
    public byte ChunkType { get; init; }
    public byte ChunkFlags { get; init; }
    public uint ChunkSize { get; init; }
    public uint Timestamp { get; init; }
    public uint? UncompressedSize { get; init; }
}

public struct QovPixel
{
    public readonly byte R;
    public readonly byte G;
    public readonly byte B;
    public readonly byte A;

    public QovPixel(byte r, byte g, byte b, byte a = 255)
    {
        R = r;
        G = g;
        B = b;
        A = a;
    }

    public static bool Equals(in QovPixel a, in QovPixel b) =>
        a.R == b.R && a.G == b.G && a.B == b.B && a.A == b.A;

    public override bool Equals(object? obj) =>
        obj is QovPixel p && Equals(in this, in p);

    public override int GetHashCode() =>
        HashCode.Combine(R, G, B, A);

    public static QovPixel operator -(in QovPixel a, in QovPixel b) =>
        new((byte)((sbyte)a.R - (sbyte)b.R), (byte)((sbyte)a.G - (sbyte)b.G), (byte)((sbyte)a.B - (sbyte)b.B), a.A);

    public static QovPixel operator +(in QovPixel a, in QovPixel b) =>
        new((byte)a.R, (byte)a.G, (byte)a.B, a.A);
}

public readonly struct QovIndexEntry
{
    public uint FrameNumber { get; init; }
    public ulong FileOffset { get; init; }
    public uint Timestamp { get; init; }
}

public struct QovFrameStats
{
    public QovHeader Header;
    public ulong FileSize;
    public int FramesDecoded;
    public uint LastTimestamp;
}

public class QovException : Exception
{
    public QovException(string message) : base(message) { }
    public QovException(string message, Exception inner) : base(message, inner) { }
}